evennia/contrib/plural.py

```diff
@@ -0,0 +1,190 @@
+
+To install, you simply need to import and inherit from the `PluralNames`
+mixin in your typeclasses.  Be sure to put it first in your inheritance
+tree.  You could do it for all your typeclasses, or just your rooms for
+instance.
+
+```python
+# Example with 'typeclasses/rooms.py'
+from evennia import DefaultRoom
+
+from evennia.contrib.plural import PluralNames
+
+class Room(PluralNames, DefaultRoom):
+
+    # ...
+```
+
+The `PluralNames` is a very light mixin that will override your
+`return_apperance` hook.  You can, however, write your own `return_appearance`
+hook and use the `group_objects` directly (see below for more information).
+
+To change the way your objects will display plural names, you can
+override the `get_plural_name` method (see the default implementation
+below).  You can do so with all your typeclasses, if you don't like the default.
+
+"""
+
+def group_objects(objects, attrname="key"):
+    """
+    Return the grouped list of objects, using their singular names.
+
+    Args:
+        objects (list ob Object): list of objects to be grouped.
+        attrname (str, optional): the name of the singular attribute.
+
+    Returns:
+        grouped (list): the same list of objects grouped by names.
+        The list will contain tuples: the number (int),
+        the object (Object) and the list of all objects of the same name.
+
+    Notes:
+        In the case of grouping (if an entry has more than 1 object),
+        only one of these objects will be displayed for grouping.
+        This is usually the expected behavior.
+
+        Grouping will occur based on the `attrname` argument.  By
+        default, the objects' `key` are used as singular names.  You
+        can provide a different attribute names.  It could be linked
+        to a property, or an Evennia attribute (just give it
+        "db.<name of attribute>" as value).
+
+    Example:
+        >>> group(rock1, tree, rock2)
+        [
+            (2, <Rock 1>, [<Rock 1>, <Rock 2>]),
+            (1, <Tree>, [<Tree>]),
+        ]
+
+    """
+    numbers = []
+    group_names = {}
+    for obj in objects:
+        # Try to get the `attrname` property
+        if "." in attrname:
+            name = obj
+            for part in attrname.split("."):
+                name = getattr(name, part)
+        else:
+            name = getattr(obj, attrname)
+
+        matches = group_names.get(name, [])
+        matches.append(obj)
+        group_names[name] = matches
+
+        # Add only if number is 1
+        if len(matches) == 1:
+            numbers.append((obj, name))
+
+    # Create the grouped list of objects
+    grouped = []
+    for obj, name in numbers:
+        matches = group_names.get(name, [])
+        number = len(matches)
+        grouped.append((number, obj, matches))
+
+    return grouped
+
+def _get_plural_name(obj, number, looker, matches):
+    """Helper function to return the plural name of an object."""
+    name = obj.key + " X " + str(number)
```

Comment:
If the session that the string is to be presented to is known, the string could be conditional on the session encoding, but just sending to the viewing character, it gets complicated in multi-session modes where there might be more than one session.